src/claimtriedb.cpp

```diff
@@ -0,0 +1,204 @@
+
+#include "claimtrie.h"
+#include "claimtriedb.h"
+
+#include <vector>
+#include <iterator>
+#include <boost/scoped_ptr.hpp>
+
+/// Concrete implementation of dirty queue
+template <typename K, typename V>
+struct CCMap : public CCBase
+{
+    CCMap() : data() {}
+    typedef std::vector<std::pair<K, V> > dataType;
+    dataType data;
+    typename dataType::iterator find(const K &key)
+    {
+        typename dataType::iterator itData = data.begin();
+        while (itData != data.end()) {
+            if (key == itData->first) break;
+            ++itData;
+        }
+        return itData;
+    }
+    void write(const size_t key, CClaimTrieDb *db)
+    {
+        for (typename dataType::iterator it = data.begin(); it != data.end(); ++it) {
+            if (it->second.empty()) {
+                db->Erase(std::make_pair(key, it->first));
+            } else {
+                db->Write(std::make_pair(key, it->first), it->second);
+            }
+        }
+        data.clear();
+    }
+};
+
+inline
+CClaimTrieDb::CClaimTrieDb(bool fMemory, bool fWipe)
+                    : CDBWrapper(GetDataDir() / "claimtrie", 100, fMemory, fWipe, false)
+{
+    doMigrate();
+}
+
+inline
+CClaimTrieDb::~CClaimTrieDb()
+{
+    for (std::map<size_t, CCBase*>::iterator itQueue = queues.begin(); itQueue != queues.end(); ++itQueue) {
+        delete itQueue->second;
+    }
+}
+
+inline
+void CClaimTrieDb::doMigrate()
+{
+    migrate<std::string, CClaimTrieNode>('n');
+    migrate<uint160, CClaimIndexElement>('i');
+    migrate<int, claimQueueRowType>('r');
+    migrate<std::string, queueNameRowType>('m');
+    migrate<int, expirationQueueRowType>('e');
+    migrate<std::string, supportMapEntryType>('s');
+    migrate<int, supportQueueRowType>('u');
+    migrate<std::string, supportQueueNameRowType>('p');
+    migrate<int, supportExpirationQueueRowType>('x');
+    Sync();
+}
+
+inline
+void CClaimTrieDb::writeQueues()
+{
+    for (std::map<size_t, CCBase*>::iterator itQueue = queues.begin(); itQueue != queues.end(); ++itQueue) {
+        itQueue->second->write(itQueue->first, this);
+    }
+}
+
+template <typename K, typename V>
+bool CClaimTrieDb::getQueueRow(const K &key, V &row) const
+{
+    const size_t hash = hashType<K, V>();
+    std::map<size_t, CCBase*>::const_iterator itQueue = queues.find(hash);
```

Comment:
As a compromise (and simplicity), perhaps it's worth considering 2 step hashes such that the above hash == some_map[K][V]?  I see why you need these now, but we're jumping through some hoops to get there, no?  What I proposed is probably less space efficient, but I want to be sure we're doing this in a sustainable way going forward.